File: src/fpl_agent/catalog.py
"""Build a normalized player/fixture catalog from bootstrap + fixtures."""

from __future__ import annotations

from typing import Any

import requests

from .api import FplApi


POSITIONS = {1: "GKP", 2: "DEF", 3: "MID", 4: "FWD"}


def _f(v: Any, default: float = 0.0) -> float:
    try:
        if v is None or v == "":
            return default
        return float(v)
    except (TypeError, ValueError):
        return default


def current_event(bootstrap: dict[str, Any]) -> dict[str, Any] | None:
    events = bootstrap.get("events") or []
    for e in events:
        if e.get("is_current"):
            return e
    for e in events:
        if e.get("is_next"):
            return e
    return events[0] if events else None


def next_event(bootstrap: dict[str, Any]) -> dict[str, Any] | None:
    events = bootstrap.get("events") or []
    for e in events:
        if e.get("is_next"):
            return e
    cur = current_event(bootstrap)
    if not cur:
        return None
    for e in events:
        if e.get("id", 0) == cur["id"] + 1:
            return e
    return None


def build_catalog(api: FplApi) -> dict[str, Any]:
    boot = api.bootstrap()
    fixtures = api.fixtures()
    teams = {t["id"]: t for t in boot.get("teams", [])}
    types = {t["id"]: t for t in boot.get("element_types", [])}
    cur = current_event(boot)
    nxt = next_event(boot)

    upcoming_by_team: dict[int, list[dict[str, Any]]] = {tid: [] for tid in teams}
    for fx in fixtures:
        if fx.get("finished") or fx.get("event") is None:
            continue
        event = fx["event"]
        if cur and event < cur["id"]:
            continue
        for side, opp, diff_key, home in (
            ("team_h", "team_a", "team_h_difficulty", True),
            ("team_a", "team_h", "team_a_difficulty", False),
        ):
            tid = fx[side]
            upcoming_by_team[tid].append(
                {
                    "event": event,
                    "opponent_id": fx[opp],
                    "opponent": teams.get(fx[opp], {}).get("short_name", "?"),
                    "difficulty": fx.get(diff_key, 3),
                    "is_home": home,
                    "kickoff_time": fx.get("kickoff_time"),
                }
            )

    for tid in upcoming_by_team:
        upcoming_by_team[tid].sort(key=lambda x: (x["event"], x.get("kickoff_time") or ""))

    players: list[dict[str, Any]] = []
    for el in boot.get("elements", []):
        if el.get("removed"):
            continue
        team_id = el["team"]
        upcoming = upcoming_by_team.get(team_id, [])[:8]
        next3 = upcoming[:3]
        next5 = upcoming[:5]
        avg_fdr_3 = sum(f["difficulty"] for f in next3) / len(next3) if next3 else 3.0
        avg_fdr_5 = sum(f["difficulty"] for f in next5) / len(next5) if next5 else 3.0

        status = el.get("status", "a")
        chance = el.get("chance_of_playing_next_round")
        if chance is None:
            chance = 100 if status == "a" else (75 if status == "d" else 0)

        xgi90 = _f(el.get("expected_goal_involvements_per_90"))
        xg90 = _f(el.get("expected_goals_per_90"))
        xa90 = _f(el.get("expected_assists_per_90"))
        ep_next = _f(el.get("ep_next"))
        ep_this = _f(el.get("ep_this"))
        form = _f(el.get("form"))
        ownership = _f(el.get("selected_by_percent"))
        cost = el.get("now_cost", 0) / 10.0
        minutes = el.get("minutes", 0) or 0
        starts = el.get("starts", 0) or 0

        players.append(
            {
                "id": el["id"],
                "web_name": el.get("web_name") or el.get("second_name"),
                "full_name": f"{el.get('first_name', '')} {el.get('second_name', '')}".strip(),
                "team_id": team_id,
                "team": teams.get(team_id, {}).get("short_name", "?"),
                "position": POSITIONS.get(el.get("element_type"), "?"),
                "position_id": el.get("element_type"),
                "cost": cost,
                "status": status,
                "news": el.get("news") or "",
                "news_added": el.get("news_added"),
                "scout_news_link": el.get("scout_news_link") or "",
                "chance_this": el.get("chance_of_playing_this_round"),
                "chance_next": chance,
                "minutes": minutes,
                "starts": starts,
                "form": form,
                "points_per_game": _f(el.get("points_per_game")),
                "total_points": el.get("total_points", 0) or 0,
                "ownership": ownership,
                "ep_this": ep_this,
                "ep_next": ep_next,
                "xg": _f(el.get("expected_goals")),
                "xa": _f(el.get("expected_assists")),
                "xgi": _f(el.get("expected_goal_involvements")),
                "xg90": xg90,
                "xa90": xa90,
                "xgi90": xgi90,
                "ict": _f(el.get("ict_index")),
                "bps": el.get("bps", 0) or 0,
                "influence": _f(el.get("influence")),
                "creativity": _f(el.get("creativity")),
                "threat": _f(el.get("threat")),
                "defensive_contribution": _f(el.get("defensive_contribution")),
                "defensive_contribution_per_90": _f(el.get("defensive_contribution_per_90")),
                "penalties_order": el.get("penalties_order"),
                "corners_order": el.get("corners_and_indirect_freekicks_order"),
                "direct_freekicks_order": el.get("direct_freekicks_order"),
                "transfers_in_event": el.get("transfers_in_event", 0) or 0,
                "transfers_out_event": el.get("transfers_out_event", 0) or 0,
                "value_form": _f(el.get("value_form")),
                "value_season": _f(el.get("value_season")),
                "avg_fdr_next3": round(avg_fdr_3, 2),
                "avg_fdr_next5": round(avg_fdr_5, 2),
                "next_fixtures": next5,
                "can_select": bool(el.get("can_select", True)),
            }
        )

    return {
        "bootstrap": boot,
        "current_event": cur,
        "next_event": nxt,
        "teams": teams,
        "types": types,
        "players": players,
        "fixtures": fixtures,
    }


def load_squad(api: FplApi, entry_id: int, catalog: dict[str, Any]) -> dict[str, Any]:
    entry = api.entry(entry_id)
    cur = catalog.get("current_event") or {}
    event_id = cur.get("id")
    if not event_id:
        raise RuntimeError("No current/next gameweek found in bootstrap data.")

    # Prefer current GW picks; if unavailable, try previous finished GW
    picks_payload: dict[str, Any] | None = None
    try:
        picks_payload = api.entry_picks(entry_id, event_id)
    except requests.HTTPError:
        picks_payload = None

    if picks_payload is None and event_id > 1:
        try:
            picks_payload = api.entry_picks(entry_id, event_id - 1)
            event_id = event_id - 1
        except requests.HTTPError as exc:
            raise RuntimeError(f"Could not load picks for entry {entry_id}") from exc

    if picks_payload is None:
        raise RuntimeError(f"Could not load picks for entry {entry_id}")

    by_id = {p["id"]: p for p in catalog["players"]}
    picks = []
    for pick in (picks_payload or {}).get("picks", []):
        base = by_id.get(pick["element"], {"id": pick["element"], "web_name": f"#{pick['element']}"})
        picks.append(
            {
                **base,
                "multiplier": pick.get("multiplier", 0),
                "is_captain": bool(pick.get("is_captain")),
                "is_vice_captain": bool(pick.get("is_vice_captain")),
                "position_slot": pick.get("position"),
                "selling_price": (pick.get("selling_price") or 0) / 10.0,
                "purchase_price": (pick.get("purchase_price") or 0) / 10.0,
            }
        )

    history = api.entry_history(entry_id)
    return {
        "entry": entry,
        "event_id": event_id,
        "picks": picks,
        "entry_history": (picks_payload or {}).get("entry_history"),
        "active_chip": (picks_payload or {}).get("active_chip"),
        "chips": history.get("chips", []),
        "current": history.get("current", []),
        "bank": (entry.get("last_deadline_bank") or 0) / 10.0,
        "team_value": (entry.get("last_deadline_value") or 0) / 10.0,
        "overall_rank": entry.get("summary_overall_rank"),
        "overall_points": entry.get("summary_overall_points"),
        "name": f"{entry.get('player_first_name', '')} {entry.get('player_last_name', '')}".strip(),
        "team_name": entry.get("name"),
    }

File: web/src/lib/fpl/catalog.ts
import {
  fetchBootstrap,
  fetchEntry,
  fetchEntryHistory,
  fetchEntryPicks,
  fetchFixtures,
} from "./api";
import type { FixtureSlice, Player, SquadData } from "./types";

const POSITIONS: Record<number, string> = {
  1: "GKP",
  2: "DEF",
  3: "MID",
  4: "FWD",
};

function num(v: unknown, fallback = 0): number {
  if (v == null || v === "") return fallback;
  const n = Number(v);
  return Number.isFinite(n) ? n : fallback;
}

type BootEl = Record<string, unknown>;
type Team = { id: number; short_name: string };
type Event = {
  id: number;
  name?: string;
  is_current?: boolean;
  is_next?: boolean;
  deadline_time?: string;
};

export function currentEvent(events: Event[]): Event | undefined {
  return events.find((e) => e.is_current) ?? events.find((e) => e.is_next) ?? events[0];
}

export function nextEvent(events: Event[]): Event | undefined {
  const nxt = events.find((e) => e.is_next);
  if (nxt) return nxt;
  const cur = currentEvent(events);
  if (!cur) return undefined;
  return events.find((e) => e.id === cur.id + 1);
}

export async function buildCatalog() {
  const [boot, fixtures] = await Promise.all([fetchBootstrap(), fetchFixtures()]);
  const events = (boot.events as Event[]) || [];
  const teamsList = (boot.teams as Team[]) || [];
  const teams = Object.fromEntries(teamsList.map((t) => [t.id, t]));
  const cur = currentEvent(events);
  const nxt = nextEvent(events);

  const upcomingByTeam: Record<number, FixtureSlice[]> = {};
  for (const t of teamsList) upcomingByTeam[t.id] = [];

  for (const fx of fixtures) {
    if (fx.finished || fx.event == null) continue;
    const event = Number(fx.event);
    if (cur && event < cur.id) continue;
    const pairs: [string, string, string, boolean][] = [
      ["team_h", "team_a", "team_h_difficulty", true],
      ["team_a", "team_h", "team_a_difficulty", false],
    ];
    for (const [side, opp, diffKey, home] of pairs) {
      const tid = Number(fx[side]);
      upcomingByTeam[tid]?.push({
        event,
        opponent_id: Number(fx[opp]),
        opponent: teams[Number(fx[opp])]?.short_name ?? "?",
        difficulty: num(fx[diffKey], 3),
        is_home: home,
        kickoff_time: (fx.kickoff_time as string) ?? null,
      });
    }
  }

  for (const tid of Object.keys(upcomingByTeam)) {
    upcomingByTeam[Number(tid)].sort((a, b) =>
      a.event !== b.event
        ? a.event - b.event
        : String(a.kickoff_time).localeCompare(String(b.kickoff_time)),
    );
  }

  const players: Player[] = [];
  for (const el of (boot.elements as BootEl[]) || []) {
    if (el.removed) continue;
    const teamId = Number(el.team);
    const upcoming = (upcomingByTeam[teamId] || []).slice(0, 8);
    const next3 = upcoming.slice(0, 3);
    const next5 = upcoming.slice(0, 5);
    const avg3 =
      next3.length > 0
        ? next3.reduce((s, f) => s + f.difficulty, 0) / next3.length
        : 3;
    const avg5 =
      next5.length > 0
        ? next5.reduce((s, f) => s + f.difficulty, 0) / next5.length
        : 3;

    const status = String(el.status ?? "a");
    let chance = el.chance_of_playing_next_round as number | null;
    if (chance == null) chance = status === "a" ? 100 : status === "d" ? 75 : 0;

    players.push({
      id: Number(el.id),
      web_name: String(el.web_name || el.second_name || ""),
      full_name: `${el.first_name ?? ""} ${el.second_name ?? ""}`.trim(),
      team_id: teamId,
      team: teams[teamId]?.short_name ?? "?",
      position: POSITIONS[Number(el.element_type)] ?? "?",
      position_id: Number(el.element_type),
      cost: num(el.now_cost) / 10,
      status,
      news: String(el.news || ""),
      news_added: (el.news_added as string) ?? null,
      scout_news_link: String(el.scout_news_link || ""),
      chance_this: (el.chance_of_playing_this_round as number) ?? null,
      chance_next: chance,
      minutes: num(el.minutes),
      starts: num(el.starts),
      form: num(el.form),
      points_per_game: num(el.points_per_game),
      total_points: num(el.total_points),
      ownership: num(el.selected_by_percent),
      ep_this: num(el.ep_this),
      ep_next: num(el.ep_next),
      xg: num(el.expected_goals),
      xa: num(el.expected_assists),
      xgi: num(el.expected_goal_involvements),
      xg90: num(el.expected_goals_per_90),
      xa90: num(el.expected_assists_per_90),
      xgi90: num(el.expected_goal_involvements_per_90),
      ict: num(el.ict_index),
      defensive_contribution_per_90: num(el.defensive_contribution_per_90),
      penalties_order: (el.penalties_order as number) ?? null,
      corners_order: (el.corners_and_indirect_freekicks_order as number) ?? null,
      direct_freekicks_order: (el.direct_freekicks_order as number) ?? null,
      avg_fdr_next3: Math.round(avg3 * 100) / 100,
      avg_fdr_next5: Math.round(avg5 * 100) / 100,
      next_fixtures: next5,
      can_select: Boolean(el.can_select ?? true),
    });
  }

  return { bootstrap: boot, current_event: cur, next_event: nxt, teams, players, fixtures };
}

export async function loadSquad(
  entryId: number,
  catalog: Awaited<ReturnType<typeof buildCatalog>>,
): Promise<SquadData> {
  const entry = await fetchEntry(entryId);
  const cur = catalog.current_event;
  if (!cur?.id) throw new Error("No current gameweek found");

  let eventId = cur.id;
  let picksPayload: Record<string, unknown> | null = null;
  try {
    picksPayload = await fetchEntryPicks(entryId, eventId);
  } catch {
    picksPayload = null;
  }
  if (!picksPayload && eventId > 1) {
    picksPayload = await fetchEntryPicks(entryId, eventId - 1);
    eventId -= 1;
  }
  if (!picksPayload) throw new Error(`Could not load picks for entry ${entryId}`);

  const byId = Object.fromEntries(catalog.players.map((p) => [p.id, p]));
  const rawPicks = (picksPayload.picks as Record<string, unknown>[]) || [];
  const picks: Player[] = rawPicks.map((pick) => {
    const base = byId[Number(pick.element)] ?? {
      id: Number(pick.element),
      web_name: `#${pick.element}`,
      full_name: "",
      team_id: 0,
      team: "?",
      position: "?",
      position_id: 0,
      cost: 0,
      status: "a",
      news: "",
      chance_next: 100,
      minutes: 0,
      starts: 0,
      form: 0,
      points_per_game: 0,
      total_points: 0,
      ownership: 0,
      ep_this: 0,
      ep_next: 0,
      xg: 0,
      xa: 0,
      xgi: 0,
      xg90: 0,
      xa90: 0,
      xgi90: 0,
      ict: 0,
      defensive_contribution_per_90: 0,
      avg_fdr_next3: 3,
      avg_fdr_next5: 3,
      next_fixtures: [],
      can_select: true,
    };
    return {
      ...base,
      multiplier: num(pick.multiplier),
      is_captain: Boolean(pick.is_captain),
      is_vice_captain: Boolean(pick.is_vice_captain),
      position_slot: num(pick.position),
      selling_price: num(pick.selling_price) / 10,
    };
  });

  const history = await fetchEntryHistory(entryId);
  return {
    entry,
    event_id: eventId,
    picks,
    chips: (history.chips as { name?: string }[]) || [],
    bank: num(entry.last_deadline_bank) / 10,
    team_value: num(entry.last_deadline_value) / 10,
    overall_rank: num(entry.summary_overall_rank) || undefined,
    overall_points: num(entry.summary_overall_points) || undefined,
    name: `${entry.player_first_name ?? ""} ${entry.player_last_name ?? ""}`.trim(),
    team_name: String(entry.name ?? ""),
  };
}
